docs/proposals/202004_embedd_cortex_frontend.md

```diff
@@ -0,0 +1,119 @@
+---
+title: Adding a New Thanos Component that Embeds Cortex Query Frontend
+type: proposal
+menu: proposals
+status: in-review
+owner: bwplotka
+---
+
+### Related Tickets
+
+* Response caching: https://github.com/thanos-io/thanos/issues/1651
+* Moving query frontend to separate repo: https://github.com/cortexproject/cortex/issues/1672
+
+## Summary
+
+This proposal describes addition of a new Thanos command (component) into `cmd/thanos` called `frontend`.
+This component will literally import a certain version of Cortex [frontend package](https://github.com/cortexproject/cortex/tree/4410bed704e7d8f63418b02b328ddb93d99fad0b/pkg/querier/frontend).
```

Comment:
Entirely practical concern, will this make https://github.com/thanos-io/thanos/issues/2415 an even bigger pain? :thinking: 

It doesn't seem like it would be a problem for Thanos, but would Cortex even still be able to import Thanos?

Reply:
Good point, so far cyclic modules are totally fine. As long as we don't introcude cyclil package dependency it should be fine :thinking: 

Anyway if we bump into any problem we will need  to reopen: https://github.com/cortexproject/cortex/issues/1672 (:

Reply:
> As long as we don't introcude cyclil package dependency it should be fine

I'm not super sure there are no cyclic package dependencies. I think the problem is caused by indirect dependencies. Which packages would Thanos directly import? `pkg/querier/frontend` and `pkg/querier/queryrange`? The last one is quite problematic, because has many internal dependencies.

Moreover, recent developments in the query-frontend have introduced the sum sharding, so we would end up importing a bunch of logic related to Cortex chunks which Thanos doesn't really need.

I would suggest to work on a refactoring and generalisation to avoid this problem. Some Cortex-specific logic should be plugged (via interfaces) only in Cortex.

Reply:
Hm, yea - this might require more work then ):

Reply:
Added to work plan.